[SPARK-10992][SQL][WIP]Partial Aggregation Support for Hive UDAF

chenghao-intel opened the pull request:
Add Hive UDAF support for aggregate2. A known issue is we still don't support the `distinct` keyword for Hive UDAF.

chenghao-intel committed:
Add Hive UDAF support for aggregate2

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39022/consoleFull) for   PR 7788 at commit [`d1f30a5`](https://github.com/apache/spark/commit/d1f30a52b91e3024f746953cf84081db317bea49).

SparkQA commented:
[Test build #39022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39022/console) for   PR 7788 at commit [`d1f30a5`](https://github.com/apache/spark/commit/d1f30a52b91e3024f746953cf84081db317bea49).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai commented:
Is it possible that we just keep `HiveUdaf2` and do not change other places? Also, can we remove `terminatePartial`? So, every time we do update or merge, we always store the updated buffer value back to the spark sql row. I think it is fine if it is not very efficient.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -632,3 +634,123 @@ private[hive] case class HiveUDAFFunction(
+      case Final => evaluator.init(GenericUDAFEvaluator.Mode.FINAL, Array(bufferObjectInspector))
+      case Complete => evaluator.init(GenericUDAFEvaluator.Mode.COMPLETE, inspectors)
+      case Partial => evaluator.init(GenericUDAFEvaluator.Mode.PARTIAL1, inspectors)
+      case x => throw new IllegalArgumentException(s"We don't support $x yet.")
```
PartialMerge will be GenericUDAFEvaluator.Mode.PARTIAL2, right?

chenghao-intel replied:
Yes, that's true, but I am not so sure if we support the 2 shuffle stages for aggregation, do we? It's  probably used with `distinct` keyword specified, but we don't support that yet in Hive UDAF.

yhuai replied:
oh, we use it for distinct. An aggregate function is not aware of distinct. The planner takes care the plan rewriting work.

chenghao-intel commented:
That's quite confusing when adding the `terminatePartial`, but the reason that I added for performance concern with Hive integration.

`terminatedPartial` will be called once after we finish the map side iterating for each group, however, if we removed the `terminatePartial` function from the UDAF interface, we have to call the hive `terminatedPartial` within the `update` implementation, that will be very expensive.

chenghao-intel commented:
Mark it as WIP, I will take more time in investigating that how to integrate with Hive UDAF.

JoshRosen commented:
@yhuai @cloud-fan @chenghao-intel, has this PR been subsumed by #8874 / https://issues.apache.org/jira/browse/SPARK-10765?

yhuai commented:
Yeah, #8874 subsumes this. 

@chenghao-intel Can you close this one? Thanks!

chenghao-intel commented:
I am going to close this one for now, however, I don't think #8874 fixes the same thing with what I did. This PR actually aims to support partial aggregation for Hive UDAF with the new UDAF interface.

JoshRosen commented:
@chenghao-intel, do you want to file a new JIRA to track only the partial aggregation enhancements for Hive UDAF?

chenghao-intel commented:
Yes @JoshRosen I've just created the issue https://issues.apache.org/jira/browse/SPARK-10992, it will be great if you assign this issue to me. :)

chenghao-intel commented:
Update the title and reopen it.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #43373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43373/consoleFull) for   PR 7788 at commit [`d1f30a5`](https://github.com/apache/spark/commit/d1f30a52b91e3024f746953cf84081db317bea49).

SparkQA commented:
[Test build #43373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43373/console) for   PR 7788 at commit [`d1f30a5`](https://github.com/apache/spark/commit/d1f30a52b91e3024f746953cf84081db317bea49).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class AggregateUtils `
  * `        case other => substitute(other) // subclass can override the function `subsitute``

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43373/
Test FAILed.

JoshRosen commented:
@chenghao-intel, thanks for filing the new JIRA. Ping me once you've updated this to resolve the merge conflicts.

yhuai commented:
@chenghao-intel how about we close this for now?

chenghao-intel commented:
ok, closing it now
